docs/proposals/05-etcd-operator-tasks.md

```diff
@@ -0,0 +1,416 @@
+
+##### Task Config
+
+We do not need any config for this task. When creating an instance of `EtcdOperatorTask` for this scenario, `.spec.config` will be set to nil (unset).
+
+##### Pre-Conditions
+
+* There should not be a `on-demand snapshot compaction` task already running for the same etcd cluster.
+
+> Note: `on-demand snapshot compaction` runs as a separate job in a separate pod, which interacts with the backup bucket and not the etcd cluster itself, hence it doesn't depend on the health of etcd cluster members.
+
+#### Trigger on-demand full/delta snapshot
+
+`Etcd` custom resource provides an ability to set [FullSnapshotSchedule](https://github.com/gardener/etcd-druid/blob/master/api/v1alpha1/types_etcd.go#L158) which currently defaults to run once in 24 hrs. [DeltaSnapshotPeriod](https://github.com/gardener/etcd-druid/blob/master/api/v1alpha1/types_etcd.go#L171) is also made configurable which defines the duration after which a delta snapshot will be taken.
+If a human operator does not wish to wait for the scheduled full/delta snapshot, they can trigger an on-demand (out-of-schedule) full/delta snapshot on the etcd cluster, which will be taken by the `leading-backup-restore`.
+
+##### Possible scenarios
+
+* An on-demand full snapshot can be triggered if scheduled snapshot fails due to any reason.
+* [Gardener Shoot Hibernation](https://github.com/gardener/gardener/blob/master/docs/usage/shoot_hibernate.md): Every etcd cluster incurs an inherent cost of preserving the volumes even when a gardener shoot control plane is scaled down, i.e the shoot is in a hibernated state. However, it is possible to save on hyperscaler costs by invoking this task to take a full snapshot before scaling down the etcd cluster, and deleting the etcd data volumes afterwards.
+* [Control Plane Migration](https://github.com/gardener/gardener/blob/master/docs/proposals/07-shoot-control-plane-migration.md): In [gardener](https://github.com/gardener/gardener), a cluster control plane can be moved from one seed cluster to another. This process currently requires the etcd data to be replicated on the target cluster, so a full snapshot of the etcd cluster in the source seed before the migration would allow for faster restoration of the etcd cluster in the target seed.
+
+##### Task Config
+
+```go
+// SnapshotType can be full or delta snapshot.
+type SnapshotType string
+
+const (
+  FullSnapshot SnapshotType = "full-snapshot"
+  DeltaSnapshot SnapshotType = "delta-snapshot"
+)
+
+type OnDemandSnapshotTaskConfig struct {
+  // Type of on-demand snapshot.
+  Type SnapshotType `json:"type"`
+}
+```
+
+```yaml
+spec:
+  config: {
+    "type": <type of on-demand snapshot>,
```

Comment:
Thanks for adding YAML representation as well. Can you please remove the double-quotes from the field keys? For this as well as other YAML representations of `config`, since the double-quotes are not necessary, even for raw extensions.

Reply:
why you think so, it's not necessary ? It's a raw extension which comes as `[]byte` later decode to `string` by controller. IMO, it's good to convey the idea again that it's a raw extension, not normal yaml.
please take a look at [here](https://github.com/kubernetes/apimachinery/blob/829ed199f4e0454344a5bc5ef7859a01ef9b8e22/pkg/runtime/types.go#L77-L80)

Reply:
Double quoting keys is necessary for JSON objects, but not for YAML. You can check this example [here](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#rawextension), or also look at `spec.provider.controlPlaneConfig` or `spec.provider.infrastructureConfig` for any Gardener shoot.

One thing I also noticed is that rawExtension cannot be any arbitrary structure, but also has an `apiVersion` and `kind`. I think this needs to be relooked at in this proposal.

cc @unmarshall

Reply:
make sense.
> but also has an apiVersion and kind

I also saw this, but don't any strong opinion on this. I can add that too.

Reply:
I have addressed your review comments @shreyas-s-rao  and I have also added `apiVersion` and `kind`. PTAL

Reply:
Thanks for making the change.